botocore/serialize.py

```diff
@@ -431,6 +433,218 @@ def _serialize_type_blob(self, serialized, value, shape, key):
+        timestamp = self._convert_timestamp_to_str(value)
+        tag = 1  # Use tag 1 for unix timestamp
+        initial_byte = self._get_initial_byte(tag_major_type, tag)
+        serialized.extend(initial_byte)  # Tagging the timestamp
+        additional_info, num_bytes = self._get_additional_info_and_num_bytes(
+            timestamp
+        )
+
+        int_major_type = 0
+        if num_bytes == 0:
+            initial_byte = self._get_initial_byte(int_major_type, timestamp)
+            serialized.extend(initial_byte)
+        else:
+            initial_byte = self._get_initial_byte(
+                int_major_type, additional_info
+            )
+            serialized.extend(
+                initial_byte + timestamp.to_bytes(num_bytes, "big")
+            )
+
+    def _serialize_type_float(self, serialized, value, shape, key):
+        if special_value_bytes := self._get_bytes_for_special_numbers(value):
+            serialized.extend(
+                special_value_bytes
+            )  # Special values like NaN or Infinity
+        else:
+            major_type = 7
+            initial_byte = self._get_initial_byte(major_type, 26)
+            serialized.extend(initial_byte + struct.pack(">f", value))
+
+    def _serialize_type_double(self, serialized, value, shape, key):
+        if special_value_bytes := self._get_bytes_for_special_numbers(value):
+            serialized.extend(
+                special_value_bytes
+            )  # Special values like NaN or Infinity
+        else:
+            major_type = 7
+            initial_byte = self._get_initial_byte(major_type, 27)
+            serialized.extend(initial_byte + struct.pack(">d", value))
+
+    def _serialize_type_boolean(self, serialized, value, shape, key):
+        major_type = 7
+        additional_info = 21 if value else 20
+        serialized.extend(self._get_initial_byte(major_type, additional_info))
+
+    def _get_additional_info_and_num_bytes(self, value):
+        # Values under 24 can be stored in the initial byte and don't need further
+        # encoding
+        if value < 24:
+            return value, 0
+        # Values between 24 and 255 (inclusive) can be stored in 1 byte and
+        # correspond to additional info 24
+        elif value < 256:
+            return 24, 1
+        # Values up to 65535 can be stored in two bytes
+        elif value < 65536:
+            return 25, 2
+        # Values up to 4294967296 can be stored in four bytes
+        elif value < 4294967296:
+            return 26, 4
+        # The maximum number of bytes currently allowed by CBOR is 8
+        else:
+            return 27, 8
+
+    def _get_initial_byte(self, major_type, additional_info):
+        # The first three bits are the major type, so we need to bitshift the major
+        # type by 5
+        major_type_bytes = major_type << 5
+        return (major_type_bytes | additional_info).to_bytes(1, "big")
+
+    def _get_bytes_for_special_numbers(self, value):
```

Comment:
So special numbers are handled the same for both floats and doubles?

Reply:
Yep :) they're the same in CBOR.  Both are major type 7, just stored in different number of bytes.  

The reason we have different methods is that the serializer is for going from C2J types to CBOR data, and they are different types in C2J.  We serialize to the proper major type based on the C2J type, so both need to be handled.  

We could theoretically consolidate some of this logic into a common method, but I can't think of a more readable way to do that than what I've written here.  It would have to either branch into an if/else based on the C2J type or do unnecessary checks.  I'm open to suggestions here.